### ISSUE TITLE
AGM Overheating breaks Advanced Balistics Turret Adjustment

### ISSUE THREAD

**Issue by brainslush** [KoffeinFlummi/AGM#1936]
The new version of the AGM Overheating breaks the turret adjustment of Advanced Balistics. The cause is probably the function AGM_Core_fnc_changeProjectileDirection which is called in line 118 of fn_overheat.sqf. It worked with the old AGM Overheating.

**Comment by esteldunedain** [comment]
>  breaks the turret adjustment of Advanced Balistics.

How does it break it? 

That function is used to randomly deviate the bullets when the weapon overheats, but the same has been done in AGM_Overheating since the beginning.

**Comment by brainslush** [comment]
Well, it was done in a different way. I'm still trying to narrow down the source of the error but in earlier versions only the velocity vector was changed. and not the bullet pointing vector or whatever you want to call it. I'll have to cross check both codes to find the reason.

**Comment by brainslush** [comment]
This should do the trick for now. I'll think of a solution which allows coexistence.
```if(!isClass(configfile >> "CfgPatches" >> "advancedballistics"))then{
 line 118
};```

**Comment by esteldunedain** [comment]
> This should do the trick for now. I'll think of a solution which allows coexistence.

Obviously that would mask the issue, but at the same time would partially cripple AGM_Overheating for anyone trying to use AB; can you describe exactly what's the problem?

**Comment by brainslush** [comment]
Both addons use the fired EH and hence both addons alter the velocity vector of the bullet in the same frame. In case of AGM it simulates the dispersion of the bullet and in case of AB to take care of the turret adjustment. In my case the turret adjustment of AB is broken. It seems that object or maybe only bullet behavior can't be altered twice in the same frame or isn't applied until the next frame.

Maybe I should look for a fix in AB and not AGM

**Comment by esteldunedain** [comment]
> Both addons use the fired EH and hence both addons alter the velocity vector of the bullet in the same frame. In case of AGM it simulates the dispersion of the bullet and in case of AB to take care of the turret adjustment. 

True.

>  It seems that object or maybe only bullet behavior can't be altered twice in the same frame or isn't applied until the next frame.

But there are no problems on doing so! In AGM we may change the velocity of a single bullet on fired EH as much as four times (e.g. overheating + scopes + wind + recoil) and that doesn't generate any issues.

That's why I think we should better pinpoint what's actually causing the issue. 

Does it raise an error?

**Comment by commy2** [comment]
>It seems that object or maybe only bullet behavior can't be altered twice in the same frame or isn't applied until the next frame.

You can alter the direction and speed multiple times, because all used scripting commands take effect immediately and so pick up the effects of previously executed eventhandlers.
Now there is room for optimization here, but thats for another update.

You don't have to explain to us how our own code works. Just state what the actual issue is.

**Comment by brainslush** [comment]
I stated what the issue is. I just want to help since you probably don't know how the AB code works.

I pin pointed the acutal problem. It is the way how you calculate the new bullet velocity in AGM_Core_fnc_changeProjectileDirection. You use the direction vector and not the old momentum vector to calculate the new velocity vector. AB only alters the momentum vector of the bullet and never the direction and up vector of the bullet.

**Comment by esteldunedain** [comment]
> I stated what the issue is.

We still don't understand clearly in what way is AB broken. Does it throw errors? The bullet trayectory is not adjusted correctly by AB?

> I pin pointed the acutal problem. It is the way how you calculate the new bullet velocity in AGM_Core_fnc_changeProjectileDirection. You use the direction vector and not the old momentum vector to calculate the new velocity vector. AB only alters the momentum vector of the bullet and never the direction and up vector of the bullet.

Ok. There's a good reason to rotate the projectile. If not, things like tracers will not be aligned to the direction of moment (which looks pretty bad). The direction of the projectiles is also important for things like rockets and such. Bottom line, using the momentum vector to infer direction would break these things.

I think the best solution would be for AB to rotate the bullet to align it to the deflected trayectory.

**Comment by brainslush** [comment]
> We still don't understand clearly in what way is AB broken. Does it throw errors? The bullet trayectory is not adjusted correctly by AB?
The turret adjustment doesn't work anymore. So I choose a certain Mildot adjustment, shoot, and the bullet hits the dirt at the same spot as without dialing any adjustments. Why, b/c AB only alters the momentum vector and not the direction vector to take care of the turret adjustment.

> Ok. There's a good reason to rotate the projectile. If not, things like tracers will not be aligned to the direction of moment (which looks pretty bad). The direction of the projectiles is also important for things like rockets and such.
I understand the reason of the projectile rotation but bullets do not always point in the direction of their momentum, for example in case of cross wind but those are redundant details and not part of the discussion.

>  Bottom line, using the momentum vector to infer direction would break these things.
I didn't suggest to use the momentum vector to determine the direction direction of the bullet. In fact I suggested nothing. You could calculate both the momentum vector and the direction vector separately without breaking the code or the functionality.

> I think the best solution would be for AB to rotate the bullet to align it to the deflected trajectory.
I don't know yet if this is an option since this "cosmetic" feature may cause unnecessary computation time which Ruthberg clearly want's to avoid in order to maintain performance.

**Comment by esteldunedain** [comment]
> You could calculate both the momentum vector and the direction vector separately without breaking the code or the functionality.

Maybe that would work, but it would require us to double the amount of calculations for every user just in case they are using AB.

As I said, the best solution would be for AB to rotate the bullet as he rotates the velocity vector. Compared to all the other things it does it's probably negligible performance-wise.

**Comment by brainslush** [comment]
> As I said, the best solution would be for AB to rotate the bullet as he rotates the velocity vector. Compared to all the other things it does it's probably negligible performance-wise.

Not really since all the heavy computations in AB are outsourced to the dll which performs many times faster than the VM of Arma so that the only things slowing down AB are the commands called in the SQF in order to handle the in- and output of the Extension. I'll correspond with Ruthberg and see what comes out.

> Maybe that would work, but it would require us to double the amount of calculations for every user just in case they are using AB.

That's indeed true but a simple if statement should hande the problem without any performance penalty. I even would provide the code but it seems that commy is still working on that function.

**Comment by esteldunedain** [comment]
> I'll correspond with Ruthberg and see what comes out.

Please do that. It's a very easy, quick and logical fix for AB. In my opinion there's absolutely no reason to patch it on our side.

**Comment by commy2** [comment]
Adjusting a projectiles UP-vector in conjunction with it's velocity vector is NOT cosmetic. It's required for the projectile to be properly synched across the network. I made that mistake once with the Puma's FCS when it was still part of the BWMod.
It seems that AGM fixes the projectiles velocity vector by rightfully assuming it to be alligned with the direction vector.
It's like the bullets from AB exit the barrel tilted, which is neither realistic nor supported by the Arma engine.

Regarding performance. The math commands have a negligible influence on performance compared to the actual object adjusting commands like `setVelocity` and `setVectorDirAndUp`, especially after the vector commands were added in A3.
